### DISCUSSION TITLE
KiDiff

### DISCUSSION THREAD

**Discussion by mdeweerd** [INTI-CMNB/KiBot#265] [General]
@set-soft I started testing KiDiff.  I think that the "Issues" are related to the KiBot setup, I am not creating an issue (as of yet), but writing down observations as I go along.

# Submodules

I use submodules to refer to my libraries which appears in .git/config as follows:
```
[submodule "LIBS/kicad-ynamics-lib"]
        url = ssh://git@SERVER:PORT/NAME/PROJECT.git
        active = true
        update = merge
```

It appears that submodules are not checked out/update with regards to the project tag.  Not an issue for me, but should at least be mentioned as a warning in the documentation.

**Comment by mdeweerd** [comment]
Got the fllowing in my verbose log:

```plaintext
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 (kibot - out_diff.py:37)
DEBUG:Restoring changes (kibot - out_diff.py:147)
DEBUG:Executing: ['/usr/bin/git', 'stash', 'pop'] (kibot - out_diff.py:41)
ERROR:Running ['/usr/bin/git', 'stash', 'pop'] returned 1 (kibot - out_diff.py:45)
DEBUG:- Output from command: No stash entries found.
```

and without verbose:

```
ERROR:Running ['/usr/bin/git', 'stash', 'pop'] returned 1 (kibot - out_diff.py:45)
```

Does not happen when a file is changed (and not committed).

**Reply by set-soft** [reply]
> Got the fllowing in my verbose log:

Thanks! I didn't know about this asymmetry: push returns 0 if nothing to push, pop returns 1 if nothing to pop ...
Now should work.

**Comment by mdeweerd** [comment]
I got this when diffing the schematic:
```plaintext
ERROR:Running ['/usr/bin/kicad-diff.py', '-vvvv', '--only_cache', '--old_file_hash', 'schd096a20d3bdbc089d735159ef0ff0e0713780496', '--cache_dir', '.cache', '/root/workdir/STAR-LCD-DUO.sch', '/root/workdir/STAR-LCD-DUO.sch'] returned 3 (kibot - out_diff.py:45)
```

Apparently, eeschema does not understand --output_name .
```
DEBUG:kicad-diff.py:Executing: ['eeschema_do', 'export', '--file_format', 'pdf', '--monochrome', '--no_frame', '--output_name', '.cache/schd096a20d3bdbc089d735159ef0ff0e0713780496/Schematic.pdf', '/root/workdir/STAR-LCD-DUO.sch', '.']
usage: eeschema_do [-h] [--record] [--rec_width REC_WIDTH]
                   [--rec_height REC_HEIGHT] [--start_x11vnc] [--use_wm]
                   [--verbose] [--version] [--wait_key]
                   [--wait_start WAIT_START] [--time_out_scale TIME_OUT_SCALE]
                   {export,run_erc,netlist,bom_xml} ... schematic output_dir
eeschema_do: error: unrecognized arguments: --output_name .cache/schd096a20d3bdbc089d735159ef0ff0e0713780496/Schematic.pdf
ERROR:kicad-diff.py:Failed to plot .cache/schd096a20d3bdbc089d735159ef0ff0e0713780496/Schematic.pdf
 (kibot - out_diff.py:37)
```

I am using the dev image up-to-date:
```plaintext
dev: Pulling from setsoft/kicad_auto
Digest: sha256:9934cc7f342c9fda9326661f85e3235c49eb6de3b0a7e510d22d623537fe7fa5
Status: Image is up to date for setsoft/kicad_auto:dev
```

EDIT: Also tried `setsoft/kicad_auto_test:latest` but does not have 'diff'.

**Reply by set-soft** [reply]
Hi @mdeweerd !
This sounds strange, KiAuto introduced `--output_name` in 2.0.0 released 20 days ago. I pulled the image a few minutes ago and it contains 2.0.2.
Please double check it.

**Reply by mdeweerd** [reply]
```raw
dev: Pulling from setsoft/kicad_auto
1efc276f4ff9: Already exists
2ed77b34ec7d: Already exists
91f0ed5b3ea9: Pull complete
8ce339d63276: Pull complete
Digest: sha256:9e88f9e1a78192723b9f9e9bd88dfb46c16d3881c57d28d073c02c80192c7a52
Status: Downloaded newer image for setsoft/kicad_auto:dev
docker.io/setsoft/kicad_auto:dev
KiBot 1.2.1 - Copyright 2018-2022, Salvador E. Tropea/INTI/John Beard - License: GPL v3+
```
I get
```
DEBUG:Executing: ['/usr/bin/git', 'checkout', '--recurse-submodules', 'master'] (kibot - out_diff.py:43)
```
so it seems to be an updated version, but I am running into another issue so I need to check why.

**Reply by set-soft** [reply]
`eeschema_do` is from KiAuto, check its version

**Reply by mdeweerd** [reply]
```
grep -ir __version__ /usr/lib/python3/dist-packages/kiauto/misc.py
__version__ = '1.5.15
```

I discovered kibot-check where I did not find "KiAuto", but I guess it is "KiCad Automation tools" which I understood afterwards - maybe add "KiAuto" on that line?
```
KiBot installation checker

Core:
Linux: 5.10.102.1 (Linux 3cb3f4068566 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 GNU/Linux)
Python: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110]
KiCad: 5.1.9+dfsg1-1
Kibot: 1.2.1

Modules:
Colorama: 0.4.4
LXML: 4.6.3
PyYAML: 5.3.1
QRCodeGen: Ok
Requests: 2.25.1
XLSXWriter: 1.1.2

Tools:
Ghostscript: 9.53.3
Git: 2.30.2 (git version 2.30.2)
ImageMagick: 6.9.11.60 (Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org)
Interactive HTML BoM: 2.5.0 (v2.5.0)
KiBoM: 1.8.0 (KiBOM Version: 1.8.0)
KiCad Automation tools: 1.5.15 (pcbnew_do 1.5.15 - Copyright 2018-2021, INTI/Productize SPRL - License: Apache)
KiCad PCB/SCH Diff: 2.4.1 (kicad-diff.py 2.4.1 - Copyright 2020-2022, INTI/Salvador E. Tropea - License:)
KiCost: 1.1.11 (KiCost v1.1.11)
Pandoc: Not available
PcbDraw: 0.9.0.3 (PcbDraw 0.9.0-3)
RAR: 5.50.0 (RAR 5.50   Copyright (c) 1993-2017 Alexander Roshal   11 Aug 2017)
RSVG tools: 2.50.3 (rsvg-convert version 2.50.3)

* KiCad Automation tools not installed or too old
  Visit: https://github.com/INTI-CMNB/KiAuto
  Download it from: https://github.com/INTI-CMNB/KiAuto/releases
  Please install the Python Installer (pip).
  - Mandatory for: `gencad`, `netlist`, `pdf_pcb_print`, `pdf_sch_print`, `render_3d`, `run_drc`, `run_erc`, `step`, `svg_pcb_print`, `svg_sch_print`, `update_xml`
  - Optional to print the page frame in GUI mode for `pcb_print`

* Pandoc not installed or too old
  Install the `pandoc` package, i.e.: `sudo apt-get install pandoc`
  You should also install the following packages: texlive-latex-base, texlive-latex-recommended
  In CI/CD environments: the `kicad_auto_test` docker image contains it.
  - Optional to create PDF/ODF/DOCX files for `report`


Color reference: ok, optional for an output, optional for general use, mandatory for an output, mandatory for general use
```

**Reply by mdeweerd** [reply]
In addition to the above: how about an option to kibot to log kibot-check to a file in the output directory so that it is automatically log which tools were (last) used to generate the outputs.  It would only be correct if all files were generated "at the same time" or without updates to the tools, but I tend to clean/move the output directory when making a release.

**Reply by set-soft** [reply]
> __version__ = '1.5.15

This is an ancient version, not the one inside the docker image. Did you run the `grep` command from inside the docker image? If not: you have 1.5.15 installed in your WSL setup, and you are getting it because this is most probably requested from your home (which I guess you are mounting when using the docker image)

> maybe add "KiAuto" on that line?

I added the logic to show it for cases like KiAuto and KiDiff.

**Reply by mdeweerd** [reply]
Possibly in the cases where kibot_check reports that some tools are too old, let kibot warn/stop (according to some option).

**Reply by mdeweerd** [reply]
Ok that was completely my fault - I tried it on an old project and the script running kibot in docker still had the line that overrided kiauto with the local version...
Sorry for that.

However, that shows that a kibot-check integrated in kibot that could warn/log/block is useful ;-).

**Reply by set-soft** [reply]
KiBot should try to download a fresh copy if you have a very old installed. If KiBot can't get a copy with the correct version should abort. The problem with diff is that KiAuto is a dependency of the dependency, but I'm adding it as a direct dependency anyways.

**Comment by mdeweerd** [comment]
So I got a result for comparing two PCB git versions, the SCH diff did not work (see above).
Good work already!

Regarding the naming of the outputfile: it can of course be hardcoded in the filename, but in order to avoid repetition, it would be nice to be able to reference the git references in the name.  Either the short hashes (independeng of how the version was specified), or the tag name (specified tag name or tag present on the version when the pcb/schematic is unmodified vs. git).

I can make separate "Issues" for all of the above or combine them except for the feature request (reference hash or tag in filename for differences).

**Reply by set-soft** [reply]
> Regarding the naming of the outputfile: it can of course be hardcoded in the filename, but in order to avoid repetition, it would be nice to be able to reference the git references in the name. Either the short hashes (independeng of how the version was specified), or the tag name (specified tag name or tag present on the version when the pcb/schematic is unmodified vs. git).

Directly modifying the output name doesn't look like a good idea. Output names must be deterministic, i.e. usable as a Makefile target. Using git hashes isn't. 
So I added a new option `add_link_id`, it doesn't change the output name, but creates a symbolic link with a name that contains the hashes. I added it to the `test_diff_git_3`, in a run I got `light_control-diff.pdf` as output and a link named `light_control-diff_7ac9eed-42bc6d7.pdf`.

**Reply by mdeweerd** [reply]
Ok, I understand for the Makefile target.  The link is an interesting idea but links do not work well on windows I suppose (especially when mixing cygwin and windows), and maybe not when zipping - but I did not check.

I would be fine with a copy of the file.   They are not enormous.

**Reply by set-soft** [reply]
Hi @mdeweerd !
I don't know much about WSL, please check if this works. According to Microsoft the best WSL practice is to copy the files to the Linux side and process them there. In this case this should work. Also: Modern Windows supports links natively, so it should translate the Linux operation to a native link. If you have a real proof that this doesn't work we could add another option to copy it.

About zipping: It made me change a couple of details. If you use ZIP you'll get a copy, if you use a tarball you'll get a link

**Reply by mdeweerd** [reply]
I tested this - there is a problem with the link itself, and another issue as well which is requires to at least change the linkage.

1. The link: in my case the link is useable from Cygwin, but not from the File Explorer (windows).  Maybe that is because I start docker from Cygwin but I would be surprised that this impact the symbolic links as other windows programs dont like them.

2. After multiple runs there would be multiple links pointing to the same file which changes, while the link name itself has the version information.  So I prefer the copy, or a fixed link from the generic name to the more specific name.  In case of the latter (link from the generic name to the specific name), `make`'s requirements on the timestamp should still be ok as the link itself is updated as well as the target of the link.

EDIT: Windows has a link feature since a very long time - not only "modern windows", but Cygwin and Windows diverge on how that is implemented.

**Reply by set-soft** [reply]
I don't fully understand the case: are you using docker AND WSL? One of them should be enough. And in both cases Cygwin seems to be redundant.
According to Microsoft you can just use WSL and run Linux commands from the command prompt, so you don't need Cygwin. They also claim you can start Windows applications even from a bash shell on WSL. 
When using WSL you shouldn't need the docker image, unless you need to run 2 different KiCad versions, and even in this case you can have more than one Linux on WSL.
Microsoft recommends copying the files to the WSL side before processing (performance issues when crossing file systems). Are you sure that a symlink created by a Linux running on the WSL environment, created on /mnt/c/ isn't translated to a valid Windows link?

BTW: When I say modern Windows I refer to the Windows NT branch, not the old Win32 (i.e. Windows 95, 98, Millenium, etc.)

**Reply by mdeweerd** [reply]
I am used to cygwin and cygwin is more transparent than WSL is - there is no need for running a container or VM either.  I tried WSL a few times, it's not very practical to use.
I will look into it again at a later time, I do it more or less once a year.  It's not only about what you can do, but also how (easy) you can do it.

As far as I understand it, Docker uses the WSL framework.  Using Docker has its advantages.

Copying - I am not convinced that that is the issue, especially when cloning from inside the container.  I am close to having a workflow that can do everything in a single run in the container.

BTW! Modern windows: as I've been working with Windows NT "branch" since 1995 a timespan of more than 25 years (and I started using windows with Windows 3.1, after ZX81, ZX Spectrum, and concurrently with Sinclair QL), Windows NT (branch) is no longer a standard for "Modern".

I installed 'strace' and examined what happens with the .pro file.
```
112   stat("/root/workdir/STAR-LCD-DUO.pro", {st_mode=S_IFREG|0644, st_size=3934, ...}) = 0
112   stat("/root/workdir/STAR-LCD-DUO.pro", {st_mode=S_IFREG|0644, st_size=3934, ...}) = 0
112   rename("/root/workdir/STAR-LCD-DUO.pro", "/root/workdir/STAR-LCD-DUO.pro-bak") = 0
112   openat(AT_FDCWD, "/root/workdir/STAR-LCD-DUO.pro", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = 5
```


Its not clear what process 112 is, but it is the process launching several comand, and one of them is:
```
112   execve("/usr/bin/eeschema_do", ["eeschema_do", "export", "--file_format", "pdf", "--monochrome", "--no_frame", "--output_name", ".cache/schd096a20d3bdbc089d73515"..., "/root/workdir/STAR-LCD-DUO.sch", "."], 0x7ffd8bbad5a0 /* 20 vars */ <unfinished ...>
```

It looks like it's not putting the original file back.

**Reply by set-soft** [reply]
> BTW! Modern windows: as I've been working with Windows NT "branch" since 1995 a timespan of more than 25 years (and I started using windows with Windows 3.1, after ZX81, ZX Spectrum, and concurrently with Sinclair QL), Windows NT (branch) is no longer a standard for "Modern".

According to Microsoft Windows NT was a change in the design of the Windows kernel, introducing the HAL and moving the Win32 stuff to an API. This is what modern Windows is. It also introduced NTFS, the filesystem that supports links, unlike VFAT that fakes links.
I also used Windows 3.1, and programmed for it, and this is when I realized it wasn't usable as an OS.
My first computer was a ZX81, just with a different name "Timex Sinclair 1000", this was the name used here. Then I moved to a Commodore 128. And finally to the PCs (staring with an XT using a NEC V20 CPU), I even developed ISA boards for this computer, doing DMA at 1 MB/s almost blocked the bus :-))))

> It looks like it's not putting the original file back.

The file is loaded to memory and then restored. I never though using text mode open could generate any problem ...

**Reply by mdeweerd** [reply]
Yeah, text mode generates more issues than you'ld expect - fixed a test case in some project one that that wasn't really testing what it should because it was writing and reading post-processed data (due to the text mode).

Still, it don't call a system modern because it's based on WinNT :-) - with windows 3.1

**Comment by set-soft** [comment]
> It appears that submodules are not checked out/update with regards to the project tag. Not an issue for me, but should at least be mentioned as a warning in the documentation.

KiBot is doing `git checkout xxxx`, I think that this should be enough. Any idea?

**Reply by mdeweerd** [reply]
It seems that `git checkout --recurse-submodules xxxx` does the trick.  Tried it and git did not "complain" about the submodules, while without the option it does.

**Reply by set-soft** [reply]
Thanks @mdeweerd ! 
Now both checksouts uses this option

**Reply by mdeweerd** [reply]
I now have a strange issue, in the docker container:
```
git checkout --recurse-submodules DUO_PROTO
```

gives
```
+ git checkout --recurse-submodules DUO_PROTO
error: Entry 'LIBS/kicad-xxx-lib/footprints/XXX.pretty/XXX_DUO_A.kicad_mod' not uptodate. Cannot merge.
error: Submodule 'LIBS/kicad-xxx-lib' could not be updated.
error: Cannot update submodule:
        LIBS/kicad-xxx-lib
```
And works find on native machine.

But local git is 2.36.1 and the docker image is 2.30.2 .

I then tried in the latest ubuntu image, installed git which is 2.34.1 and tried the same commands - it works.
So it looks like this is related to the git version.

**Reply by set-soft** [reply]
Never got such an error. I'm using 2.30.2. I think newer git versions are more cleaver, the possible reasons for your problem are explained here:

https://stackoverflow.com/questions/1248029/git-pull-error-entry-foo-not-uptodate-cannot-merge

One possible problem: A mess of files that just differ in the case style, they are different from Linux and the same for Windows.

**Reply by mdeweerd** [reply]
I defined the following DockerFile to upgrade git which resolves the issue:
```
FROM setsoft/kicad_auto:dev

RUN \
    echo "deb http://deb.debian.org/debian bullseye-backports main contrib non-free" >> /etc/apt/sources.list.d/backports.list \
    && apt-get update -o Dir::Etc::sourcelist="sources.list.d/backports.list" \
       -o Dir::Etc::sourceparts="-" -o APT::Get::List-Cleanup="0" \
    && apt install -y git -t bullseye-backports
```

This adds and fetches the backports list only, then upgrades git.

**Reply by mdeweerd** [reply]
I had a few succefull runs with that fix, but I've been stuck again.
The behavior is similar to the above.

Prior to upgrading the git version, the following always failed in docker:
```
git checkout --recurse-submodules DUO_PROTO
```

Now this fails when run from within kibot.

I also started a docker run interactively to a bash shell, then ran ` kibot -c STAR-LCD-DUO.kibot.yaml -b STAR-LCD-DUO.kicad_pcb -e STAR-LCD-DUO.sch -g variant=proto -vvvv diff_A_sch` and that was aborted due to the checkout error on `git checkout --recurse-submodules DUO_PROTO` (as before).  
Then I had access to the shell again and I did `git checkout --recurse-submodules DUO_PROTO` which was executed without errror (same container execution).

From the log, the git commands performed by kibot are:
```
'git' '--version'
'git' 'stash' 'push' '-m' 'KiBot_Changes_Entry'
'git' 'rev-parse' '--abbrev-ref' 'HEAD'
'git' 'checkout' '--recurse-submodules' 'DUO_PROTO'
'git' 'stash' 'list' 'stash@{0}'
```

Executing that in a fresh docker container from within a script with the following executes just fine:
```
'git' '--version'
'git' 'stash' 'push' '-m' 'KiBot_Changes_Entry'
'git' 'rev-parse' '--abbrev-ref' 'HEAD'
'git' 'checkout' '--recurse-submodules' 'DUO_PROTO'
'git' 'stash' 'list' 'stash@{0}'
# Added manually in test script - kibot does not return to the base version on Abort.
'git' 'checkout' '--recurse-submodules' 'master'
```

Then I thought about why it worked so many times - and I decided to add an empty line to my Makefile so that something would be stashed.  That worked better.  It resulted in:

```
    os.symlink(os.path.basename(name), name_comps[0]+'_'+gh1+'-'+gh2+name_comps[1])
FileExistsError: [Errno 17] File exists: 'STAR-LCD-DUO_01_20211217-diff.pdf' -> '/root/workdir/STAR-LCD-DUO_01_20211217__proto/STAR-LCD-DUO_01_20211217-diff_e769c13-None.pdf'
```
which is another issue, showing that the link is not overwritten.

I am still puzzled - is kibot not showing all the executed git commands, or is there a difference between the logged one and the executed on, or ...?

**Reply by set-soft** [reply]
> I defined the following DockerFile to upgrade git which resolves the issue:

Using the backport isn't a good option. The backport for KiCad 5 is KiCad 6, so it could produce problems.
The `git` issue you have seems to be a problem in your repo (most probably generated by using Windows), it must be fixed in the repo.
I'll add `git` from backports only for KiCad 6.

**Reply by mdeweerd** [reply]
Regarding the git issue: the current status (with the backport git) is that when I perform the git commands found in kibot's debug trace inside docker, there is no issue.  But when kibot runs them, there the issue appears.

To go a bit further, I cloned my repository using docker's git and then execute kibot in the same docker setup, and that works, but now it also works with the original setup (the original clone made outside docker).

On the second run, I got:
```
FileExistsError: [Errno 17] File exists: 'STAR-LCD-DUO_01_20211217-_sch.pdf' -> '/root/workdir/STAR-LCD-DUO_01_20211217__proto/STAR-LCD-DUO_01_20211217-_sch_e769c13-None.pdf'
```

Which is something else.

**Reply by set-soft** [reply]
> Which is something else.

This should be fixed by d834ba507575d2abaaf0279c21e4cd77fbd09981

**Reply by set-soft** [reply]
> Regarding the git issue: the current status (with the backport git) is that when I perform the git commands found in kibot's debug trace inside docker, there is no issue. But when kibot runs them, there the issue appears.
> 
> To go a bit further, I cloned my repository using docker's git and then execute kibot in the same docker setup, and that works, but now it also works with the original setup (the original clone made outside docker).

Are you sure you don't have more than one git command? KiBot explicitly runs the full qualified binary, i.e. `/usr/bin/git` and not just `git`, so you know which binary is running.

BTW: All these problems points to a problem in the repo.

**Reply by mdeweerd** [reply]
> Are you sure you don't have more than one git command?

I using gti from setsoft/kicad_auto:dev for this test.  I just ran `which git` inside that container and it returns /usr/bin/git .
All git commands are done using setsoft/kicad_auto:dev.  I have a script that first clones the repo using that git and an OAUTH key over https.

```shell
#!/bin/bash
./r /usr/bin/git clone --recurse-submodules https://oauth2:SECRET_KEY@repo_path.git starlcd
cd starlcd
make
```

Where './r' is a script to launch the command inside the docker container.  Make then runs kibot which executes in the same container.

I get:
```
ERROR:Running ['/usr/bin/git', 'checkout', '--recurse-submodules', 'master'] returned 1 (kibot - out_diff.py:50)
DEBUG:- Output from command: error: Your local changes to the following files would be overwritten by checkout:
        STAR-LCD-DUO.pro
Please commit your changes or stash them before you switch branches.
Aborting
 (kibot - out_diff.py:42)
```
When checking the git status after that run:
```
$ git branch
* (HEAD detached at DUO_PROTO)
  master
```

The differences of STAR-LCD-DUO.pro seen in the containers' git are only the line endings.  After doing unix2dos on the file, there is no longer any difference on that file.

**Reply by set-soft** [reply]
Hmmm ... looks like the project file was modified by KiCad and restored by KiAuto, but using text mode ...
I'll release a KiAuto that fixes it.

**Reply by set-soft** [reply]
Done, v2.0.3 should fix it. Docker images should be automatically generated in a few minutes.

**Reply by mdeweerd** [reply]
More than a few minutes, but it arrived and works when cloning everything in the container.

Still issues when using the clone made outside docker - I may find the real reason at some point.
It is currently failing because a file is missing in the submodules on recursive checkout...

I do not have changes in my submodules, but to be complete they would have to be stashed as well... (`git submodule foreach 'git stash'` )

**Reply by mdeweerd** [reply]
According to my tests, `git checkout -f --recurse-submodules REFERENCE` is the solution to the blocking checkout.  However, the submodules end up detached.  So  this requires to check (and store) the HEAD of each submodule before any action.

`git submodule foreach `git stash -m NAME`` works, but to pop them, one seems to need to enter ech submodule `( cd "submoduledir" ; git checkout REFERENCE)`  after checking ``git submodule foreach `git stash list'`

**Reply by set-soft** [reply]
a4abb65ef397db439e388f58583e3c1fb1ca370d should add this.
Note that recursive submodules aren't supported.

The `-f` will go as an option.

**Reply by set-soft** [reply]
e34d4a01e05c83d4cd0822cc0d03816920da0ffd adds the `force_checkout` option to add `--force`

**Comment by mdeweerd** [comment]
Given that the git issue and my local kiauto issue are now fixed, I could procede with the flow to generated difference files for the PCB and the SCH, but the SCH generated the PCB difference.

Further I am questioning myself how the difference filename is going to reflect PCB or SCH through my global naming scheme.

Here are the essential parts of my yaml configuration, I can share more if needed - I was using the proto variant, but I do not think that is essential:
```yaml
import:
   # Some imports
global:
  output: '%f_%r_%d-%i.%x'
  # out_dir: "%f_%br_%bd_%g/"
  out_dir: '%bf_%r_%d_%g/'
  date_time_format: '%Y%m%d_%H%M%S'
  date_format: '%Y%m%d'
  time_format: '%H%M%S'
  variant: 5V_3btn
variants:
  - name: proto
    comment: 5V power, 4 buttons per screen
    type: kicost
    file_id: _proto
    variant: ^(button3|button4|screen5v|conn3|AD8|bl_buffer)$
    pre_transform: kicost_rename
  - name: diff_A_pcb
    type: diff
    comment: PCB difference with previous version
    layers: [all]
    options:
      # output: '%f-%i%I%v.%x'
      cache_dir: .cache
      diff_mode: red_green # or stats
      # fuzz: 5
      #new: AYAC100577-02_20191209
      #new_type: git  # git or file
      old: DUO_PROTO
      old_type: git
      pcb: true
      threshold: 40
  - name: diff_A_sch
    type: diff
    comment: SCH difference with previous version
    layers: [all]
    options:
      # output: '%f-%i%I%v.%x'
      cache_dir: .cache
      diff_mode: red_green # or stats
      # fuzz: 5
      #new: AYAC100577-02_20191209
      #new_type: git  # git or file
      old: DUO_PROTO
      old_type: git
      pcb: false
      threshold: 40
```

First I generated both diff_A_pcb and diff_A_sch, but noticing that I got the PCB differences, I then deleted the difference PDF that was generated and only requested to produce diff_A_sch - I still got the PCB difference.
The log suggests that the Schematic difference is generated.

**Reply by set-soft** [reply]
> Given that the git issue and my local kiauto issue are now fixed, I could procede with the flow to generated difference files for the PCB and the SCH, but the SCH generated the PCB difference.

You are right, they must be identified differently. Now %i is diff_pcb for PCBs and diff_sch for Schematics

**Comment by mdeweerd** [comment]
Side effects of stash without local changes, and old stash on stack.
Kibot should only stash pop when it did really stash something.

```ansi
^[[36m^[[1mDEBUG:Executing: ['/usr/bin/git', 'stash', 'push', '-m', 'KiBot_Changes_Entry'] (kibot - out_diff.py:46)^[[0m
^[[36m^[[1mDEBUG:- Output from command: No local changes to save
```

```
^[[36m^[[1mDEBUG:Executing: ['/usr/bin/git', 'stash', 'pop', '--index'] (kibot - out_diff.py:46)^[[0m
^[[31m^[[1mERROR:Running ['/usr/bin/git', 'stash', 'pop', '--index'] returned 1 (kibot - out_diff.py:50)^[[0m
^[[36m^[[1mDEBUG:- Output from command: Auto-merging runInKiAuto
CONFLICT (content): Merge conflict in runInKiAuto
```

**Reply by set-soft** [reply]
This is already solved. KiBot only pops when the top of the stack has a magic message

**Comment by mdeweerd** [comment]
I also suggest to add '--' in the calls to git as the tags might start with the - sign and be interpreted as an option (not my case, but could happen).

**Reply by set-soft** [reply]
Doesn't seem to be possible, things like: `/usr/bin/git checkout --recurse-submodules -- HEAD` doesn't work

**Comment by mdeweerd** [comment]
My offer to create issues instead still stands - it maybe more efficient at this point.

**Reply by set-soft** [reply]
Do it, but only for things we will solve.
Various things you mention are already solved or must be solved outside KiBot.

**Reply by mdeweerd** [reply]
Actually it all seems to be fixed in KiBot now (at least it works with the project git clone I have been using).
And this is "new" which happens when creating the difference twice:
```
FileExistsError: [Errno 17] File exists: 'STAR-LCD-DUO_01_20211217-_sch.pdf' -> '/root/workdir/STAR-LCD-DUO_01_20211217__proto/STAR-LCD-DUO_01_20211217-_sch_e769c13-None.pdf'
```

**Reply by mdeweerd** [reply]
And other issue: still getting pcb differences for the schematic.  

Git is giving trouble again, also when the clone was managed in docker.  Strange.

When starting from a clean clone using docker:
```
ERROR:Running ['/usr/bin/git', 'checkout', '--recurse-submodules', 'master'] returned 1 (kibot - out_diff.py:50)
DEBUG:- Output from command: error: Your local changes to the following files would be overwritten by checkout:
        STAR-LCD-DUO.pro
```

**Reply by set-soft** [reply]
Again: Double check you don't have a mess with the upper/lower case names. 
Try copying the repo to the Linux filesystem, not mounting the NTFS

**Comment by mdeweerd** [comment]
In case of failure one out of two times I do not get into the original git state in my clone.  So I starting investigating mounting an overlay over the docker workdir and write the output elsewhere.

**Reply by set-soft** [reply]
KiBot tries to restore the old state, even after crashing (2 layers of `finally` statements are implemented). But if you get a fail during the restore process (double fault) things won't be restored.
We need to find which situations can generate such a problem and find ways to workaround them.
